### PR TITLE
New Address functions

### PR DESCRIPTION
**Pull request by dcousens**
Related to https://github.com/bitcoinjs/bitcoinjs-lib/issues/787, chained on https://github.com/bitcoinjs/bitcoinjs-lib/pull/852,  so ignore the commit from that

I don't know if this is the *best* way... it seems really messy simply because you're writing out a damn 5+ word sentence each time you want to do anything.

Any suggestions?
Could/should they merge with `bitcoin.script` at `4.0.0`?

### CODE REVIEW

**Comment by dcousens** [comment]
@junderw

File: test/integration/addresses.js [hunk 1]
```diff
@@ -49,6 +47,56 @@ describe('bitcoinjs-lib (addresses)', function () {
...
+    assert.strictEqual(address, 'bc1qt97wqg464zrhnx23upykca5annqvwkwujjglky')
+  })
+
+  it('can generate a SegWit 2-of-2 multisig address (via P2SH)', function () {
+    var pubKeys = [
+      '026477115981fe981a6918a6297d9803c4dc04f328f22041bedff886bbc2962e01',
+      '02c96db2302d19b43d4c69368babace7854cc84eb9e061cde51cfa77ca4a22b8b9'
+    ].map(function (hex) { return Buffer.from(hex, 'hex') })
+
+    var redeemScript = bitcoin.script.multisig.output.encode(2, pubKeys) // 2 of 2
```
Comment (junderw) [inline test/integration/addresses.js]: This redeemScript was fixed in the other PR but not this one.

Perhaps you rebased on the wrong branch?

File: src/address.js [hunk 2]
```diff
@@ -1,11 +1,49 @@
...
 var typeforce = require('typeforce')
 var types = require('./types')
 
+// TODO: refactor
+function p2pkhFromHash (hash, network) {
+  network = network || networks.bitcoin
+  return toBase58Check(hash, network.pubKeyHash)
+}
+
+function p2pkhFromKeyPair (keyPair) {
```
Comment (junderw) [inline src/address.js]: Would be nice if all the FromKeyPair functions could be called from the ECPair prototype.

keyPair.getAddress('p2wpkh') or maybe just a separate function?

Comment (dcousens) [inline src/address.js]: @junderw maybe pitch in your ideas to https://github.com/bitcoinjs/bitcoinjs-lib/issues/787?

**Comment by afk11** [comment]
Needs rebase

**Comment by dcousens** [comment]
Rebased,  but,  with `4.0.0` so close, I don't know if I want to go through with supporting these in this version?

Unsure

**Comment by dcousens** [comment]
With the API I am toying with in https://github.com/bitcoinjs/bitcoinjs-playground/blob/master/scripts.js,  anyone who adopts these would run an unnecessary 2nd refactor to upgrade to `4.0.0`,  after upgrading to these... ?

Maybe.

**Comment by dcousens** [comment]
Closing in favour of the API being developed in https://github.com/bitcoinjs/bitcoinjs-playground/blob/master/scripts.js

**Pull request closed without merging**